clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java

```diff
@@ -47,32 +185,166 @@ public boolean isValid() {
-        long start = time.milliseconds();
-        long elapsed = 0;
-        while (!hasPid() && elapsed < maxWaitTimeMs) {
-            wait(maxWaitTimeMs);
-            elapsed = time.milliseconds() - start;
+    public boolean isFenced() {
+        return isFenced;
+    }
+
+    public void beginTransaction() {
+        isInTransaction = true;
+    }
+
+    public boolean isCompletingTransaction() {
+        return isInTransaction && isCompletingTransaction;
+    }
+
+    public synchronized FutureTransactionalResult beginCommittingTransaction() {
+        return beginCompletingTransaction(true);
+    }
+
+    public synchronized FutureTransactionalResult beginAbortingTransaction() {
+        return beginCompletingTransaction(false);
+    }
+
+    private FutureTransactionalResult beginCompletingTransaction(boolean isCommit) {
+        if (!isCompletingTransaction) {
+            TransactionalRequestResult result = new TransactionalRequestResult();
+            FutureTransactionalResult resultFuture = new FutureTransactionalResult(result);
+
+            isCompletingTransaction = true;
+            if (!newPartitionsToBeAddedToTransaction.isEmpty()) {
+                pendingTransactionalRequests.add(addPartitionsToTransactionRequest(false));
+            }
+            pendingTransactionalRequests.add(endTxnRequest(isCommit, false, result));
+            return resultFuture;
         }
-        return pidAndEpoch;
+        return null;
+    }
+
+    public synchronized FutureTransactionalResult sendOffsetsToTransaction(Map<TopicPartition, OffsetAndMetadata> offsets,
+                                         String consumerGroupId) {
+        TransactionalRequestResult result = new TransactionalRequestResult();
+        FutureTransactionalResult resultFuture = new FutureTransactionalResult(result);
+        pendingTransactionalRequests.add(addOffsetsToTxnRequest(offsets, consumerGroupId, false, result));
+        return resultFuture;
+    }
+
+    public boolean isInTransaction() {
+        return isTransactional() && isInTransaction;
+    }
+
+
+    public synchronized void maybeAddPartitionToTransaction(TopicPartition topicPartition) {
+        if (partitionsInTransaction.contains(topicPartition))
+            return;
+        newPartitionsToBeAddedToTransaction.add(topicPartition);
+    }
+
+    public void needsRetry(TransactionalRequest request) {
+        request.setRetry();
+        pendingTransactionalRequests.add(request);
+    }
+
+    public void didNotSend(TransactionalRequest request) {
+        pendingTransactionalRequests.add(request);
+    }
+
+    public Node coordinator(FindCoordinatorRequest.CoordinatorType type) {
+        switch (type) {
+            case GROUP:
+                return consumerGroupCoordinator;
+            case TRANSACTION:
+                return transactionCoordinator;
+            default:
+                return null;
```

Comment:
It sucks that Java doesn't have exhaustiveness warnings for enums. Given that, it seems that throwing an assertion error is the best bet for impossible paths like this.